File: src/studio/backend/data/metrics.py
"""Shared metric rules: default performance metric, loss-like metrics and the experiment catalog."""

import re
from dataclasses import dataclass
from typing import Any

from .records import ExperimentRecord

TEST_TABLE = "test"
TRAINING_DATA_TABLE = "training_data"
DEFAULT_METRIC_PREFERENCE = ("score", "score-0")
LOSS_RE = re.compile(r"loss|td-error|grad", re.IGNORECASE)


@dataclass(frozen=True)
class MetricRef:
    table: str
    metric: str

    def to_json(self) -> dict[str, str]:
        return {"table": self.table, "metric": self.metric}


@dataclass
class CatalogTable:
    metrics: list[str]
    x_columns: list[str]
    runs: list[str]


@dataclass
class Catalog:
    tables: dict[str, CatalogTable]
    default_metric: MetricRef | None
    loss_metrics: list[MetricRef]

    def to_json(self) -> dict[str, Any]:
        """The API's `Catalog`. @ai-generated"""
        return {
            "tables": {name: {"metrics": t.metrics, "x_columns": t.x_columns, "runs": t.runs} for name, t in self.tables.items()},
            "default_metric": self.default_metric.to_json() if self.default_metric else None,
            "loss_metrics": [m.to_json() for m in self.loss_metrics],
        }


def is_loss_metric(name: str) -> bool:
    """@ai-generated"""
    return LOSS_RE.search(name) is not None


def default_metric(tables: dict[str, list[str]]) -> MetricRef | None:
    """
    Default performance metric: in the `test` table, `score`, else `score-0`, else the first
    numeric column alphabetically. None if there is no (non-empty) `test` table.

    @ai-generated
    """
    metrics = tables.get(TEST_TABLE)
    if not metrics:
        return None
    for preferred in DEFAULT_METRIC_PREFERENCE:
        if preferred in metrics:
            return MetricRef(TEST_TABLE, preferred)
    return MetricRef(TEST_TABLE, min(metrics))


def build_catalog(record: ExperimentRecord) -> Catalog:
    """
    Union of the runs' tables: plottable (numeric/bool, non-x) columns in first-seen order, x columns,
    and the runs providing each table. Tables are sorted by name.

    @ai-generated
    """
    tables = dict[str, CatalogTable]()
    for run in record.runs:
        for name, info in run.tables.items():
            entry = tables.setdefault(name, CatalogTable([], [], []))
            entry.metrics.extend(m for m in info.plottable if m not in entry.metrics)
            entry.x_columns.extend(x for x in info.x_columns if x not in entry.x_columns)
            entry.runs.append(run.id)
    tables = dict(sorted(tables.items()))
    loss = (
        [MetricRef(TRAINING_DATA_TABLE, m) for m in tables[TRAINING_DATA_TABLE].metrics if is_loss_metric(m)]
        if TRAINING_DATA_TABLE in tables
        else []
    )
    return Catalog(tables, default_metric({n: t.metrics for n, t in tables.items()}), loss)

File: src/studio/frontend/src/domain/metrics.ts
/**
 * Metric rules shared by presets, library sparklines and the performance timeline.
 */
import type { Catalog, MetricRef } from "../api/schemas";

export const X_COLUMNS = new Set(["time_step", "timestamp_sec"]);
export const LOSS_PATTERN = /loss|td-error|grad/;
const TABLE_ORDER = ["test", "train", "training_data"];
const TABLE_LABEL: Record<string, string> = { test: "Test", train: "Train", training_data: "Training data" };

/**
 * Default performance metric of an experiment: the catalog's `default_metric` when present,
 * otherwise the same rule as the backend on the `test` table: `score` → `score-0` → first metric
 * alphabetically (excluding x columns).
 *
 * @ai-generated
 */
export function defaultMetric(catalog: Catalog | null | undefined): MetricRef | null {
  if (!catalog) return null;
  if (catalog.default_metric) return catalog.default_metric;
  const test = catalog.tables.test;
  if (!test) return null;
  const metrics = test.metrics.filter((m) => !X_COLUMNS.has(m));
  for (const preferred of ["score", "score-0"]) if (metrics.includes(preferred)) return { table: "test", metric: preferred };
  const first = [...metrics].sort()[0];
  return first ? { table: "test", metric: first } : null;
}

/** Loss-like metrics: the catalog's `loss_metrics`, else `training_data` columns matching the pattern. @ai-generated */
export function lossMetrics(catalog: Catalog | null | undefined): MetricRef[] {
  if (!catalog) return [];
  if (catalog.loss_metrics.length) return catalog.loss_metrics;
  return (catalog.tables.training_data?.metrics ?? []).filter((m) => LOSS_PATTERN.test(m)).map((metric) => ({ table: "training_data", metric }));
}

export function hasMetric(catalog: Catalog | null | undefined, table: string, metric: string): boolean {
  return !!catalog?.tables[table]?.metrics.includes(metric);
}

/** Order tables as Test, Train, Training data, then the others alphabetically. @ai-generated */
export function sortTables(tables: Iterable<string>): string[] {
  const rank = (t: string) => {
    const i = TABLE_ORDER.indexOf(t);
    return i < 0 ? TABLE_ORDER.length : i;
  };
  return [...tables].sort((a, b) => rank(a) - rank(b) || a.localeCompare(b));
}

export const tableLabel = (t: string): string => TABLE_LABEL[t] ?? t;

/** Test-like tables (their points open the episodes sheet). */
export const isTestTable = (t: string): boolean => t.startsWith("test");
